exchange-wrapper/contracts/ExchangeWrapper.sol

```diff
@@ -5,122 +5,221 @@ pragma abicoder v2;
+        if (purchaseDetails.marketId == Markets.SeaPortAdvancedOrders){
+            (bool success,) = address(seaPort).call{value : paymentAmount}(purchaseDetails.data);
+            if (allowFail) {
+                if (!success) {
+                    return (false, 0, 0);
+                }
+            } else {
+                require(success, "Purchase SeaPortAdvancedOrders failed");
+            }
+        } else if (purchaseDetails.marketId == Markets.WyvernExchange) {
             (bool success,) = address(wyvernExchange).call{value : paymentAmount}(purchaseDetails.data);
-            require(success, "Purchase failed");
+            if (allowFail) {
+                if (!success) {
+                    return (false, 0, 0);
+                }
+            } else {
+                require(success, "Purchase wyvernExchange failed");
+            }
         } else if (purchaseDetails.marketId == Markets.ExchangeV2) {
-            (LibOrder.Order memory sellOrder, bytes memory sellOrderSignature, uint purchaseAmount) = abi.decode(purchaseDetails.data, (LibOrder.Order, bytes, uint));
-            matchExchangeV2(sellOrder, sellOrderSignature, paymentAmount, purchaseAmount);
+            (bool success,) = address(exchangeV2).call{value : paymentAmount}(purchaseDetails.data);
+            if (allowFail) {
+                if (!success) {
+                    return (false, 0, 0);
+                }
+            } else {
+                require(success, "Purchase rarible failed");
+            }
+        } else if (purchaseDetails.marketId == Markets.X2Y2) {
+            Ix2y2.RunInput memory input = abi.decode(purchaseDetails.data, (Ix2y2.RunInput));
+
+            if (allowFail) {
+                try Ix2y2(x2y2).run{value : paymentAmount}(input) {
+                } catch {
+                    return (false, 0, 0);
+                }
+            } else {
+                Ix2y2(x2y2).run{value : paymentAmount}(input);
+            }
+            for (uint i = 0; i < input.orders.length; i++) {
+                for (uint j = 0; j < input.orders[i].items.length; j++) {
+                    Ix2y2.Pair[] memory pairs = abi.decode(input.orders[i].items[j].data, (Ix2y2.Pair[]));
+                    for (uint256 k = 0; k < pairs.length; k++) {
+                        Ix2y2.Pair memory p = pairs[k];
+                        IERC721Upgradeable(address(p.token)).safeTransferFrom(address(this), _msgSender(), p.tokenId);
+                    }    
+                }
+            } 
+        } else if (purchaseDetails.marketId == Markets.LooksRareOrders) {
+            (LibLooksRare.TakerOrder memory takerOrder, LibLooksRare.MakerOrder memory makerOrder, bytes4 typeNft) = abi.decode(purchaseDetails.data, (LibLooksRare.TakerOrder, LibLooksRare.MakerOrder, bytes4));
+            if (allowFail) {
+                try ILooksRare(looksRare).matchAskWithTakerBidUsingETHAndWETH{value : paymentAmount}(takerOrder, makerOrder) {
```

Comment:
do you remember why for LooksRare  we use method **matchAskWithTakerBidUsingETHAndWETH**
but for X2Y2  Ix2y2(x2y2).run used?

Reply:
maybe for LooksRare we also need ILooksRare(looksRare).run ?
need to estimate gas